solang-parser/src/helpers/fmt.rs

```diff
@@ -0,0 +1,2518 @@
+                Ok(())
+            }
+            Self::NumberLiteral(_, value, exponent, ident) => {
+                f.write_str(value)?;
+                if !exponent.is_empty() {
+                    f.write_char('e')?;
+                    f.write_str(exponent)?;
+                }
+                write_opt!(f, ": ", ident);
+                Ok(())
+            }
+            Self::HexNumberLiteral(_, value, ident) => {
+                f.write_str(value)?;
+                write_opt!(f, ": ", ident);
+                Ok(())
+            }
+            Self::HexStringLiteral(value, ident) => {
+                value.fmt(f)?;
+                write_opt!(f, ": ", ident);
+                Ok(())
+            }
+            Self::StringLiteral(value, ident) => {
+                value.fmt(f)?;
+                write_opt!(f, ": ", ident);
+                Ok(())
+            }
+            Self::Variable(ident) => ident.fmt(f),
+            Self::FunctionCall(call) => call.fmt(f),
+            Self::SuffixAccess(_, l, r) => {
+                l.fmt(f)?;
+                f.write_char('.')?;
+                r.fmt(f)
+            }
+        }
+    }
+}
+
+impl Display for pt::YulStatement {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            Self::Block(inner) => inner.fmt(f),
+            Self::FunctionDefinition(inner) => inner.fmt(f),
+            Self::FunctionCall(inner) => inner.fmt(f),
+            Self::For(inner) => inner.fmt(f),
+            Self::Switch(inner) => inner.fmt(f),
+
+            Self::Assign(_, exprs, eq_expr) => {
+                write_separated(exprs, f, ", ")?;
+                f.write_str(" := ")?;
+                eq_expr.fmt(f)
+            }
+            Self::VariableDeclaration(_, vars, eq_expr) => {
+                f.write_str("let")?;
+                if !vars.is_empty() {
+                    f.write_char(' ')?;
+                    write_separated(vars, f, ", ")?;
+                }
+                write_opt!(f, " := ", eq_expr);
+                Ok(())
+            }
+
+            Self::If(_, expr, block) => {
+                f.write_str("if ")?;
+                expr.fmt(f)?;
+                f.write_char(' ')?;
+                block.fmt(f)
+            }
+
+            Self::Leave(_) => f.write_str("leave"),
+            Self::Break(_) => f.write_str("break"),
+            Self::Continue(_) => f.write_str("continue"),
+
+            Self::Error(_) => Ok(()),
+        }
+    }
+}
+
+impl Display for pt::YulSwitchOptions {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            Self::Case(_, expr, block) => {
+                f.write_str("case ")?;
+                expr.fmt(f)?;
+                f.write_str(" ")?;
+                block.fmt(f)
+            }
+            Self::Default(_, block) => {
+                f.write_str("default ")?;
+                block.fmt(f)
+            }
+        }
+    }
+}
+
+#[inline]
```

Comment:
Can you elaborate here why we need this? It is is almost always unnecessary and deserves  a comment that proves this is needed here.

Reply:
very simple functions that only call another function, I don't see why not to add this; same for all the helper functions that do simple matching or just a field access

Reply:
This function looks a lot like it should be getting inlined by the compiler automatically, especially considering it's private. If I see this `#[inline]` here I'd assume that the compiler doesn't inline this function when it should
and it is used in a hot path somewhere where removing it introduces significant performance penalties.

If these assumptions are correct, it would be nice to leave a comment explaining that, otherwise the `#[inline]` is probably unnecessary.

Reply:
The compiler knows perfectly well when things should be inlined or not. It does not sound there is a solid reason for the `#[inline]` attribute.

Reply:
> The compiler knows perfectly well when things should be inlined or not. It does not sound there is a solid reason for the `#[inline]` attribute.

I don't believe that is the case. You are probably right in saying that since these functions are private then the compiler is more likely to inline them, but giving extra hints that they should be doesn't hurt, especially because they are private and that they should probably be inlined regardless.

> If these assumptions are correct, it would be nice to leave a comment explaining that (ideally pointing to some benchmarks or tests), otherwise the `#[inline]` is probably unnecessary.

I don't have benchmarks nor do I intend to write any simply to justify a mere compiler hint that may or may not sway performance and/or compilation speed by less than a single percentage point. Since there are almost no downsides to adding these hints I would rather just leave them as is.

This being said, if you still want me to remove the `#[inline]` hints on these private functions, I will do so.

Reply:
> They are a maintenance burden, especially without any comment providing context.

Given:
- `fn a<T>(x: T, y, z) { ... }`
- `#[inline] fn b(x: &[u8]) { a(x, <default_y>, <default_z>) }`

Since `b` is just a simple shortcut for a special case of `a`, is the reason to add `#[inline]` not obvious, or at least deductible, without requiring 2 paragraphs of comments detailing why this should be the case?

I could've just not written `b` and instead write manual calls to `a` everywhere, but since I noticed a common pattern for the same type I wrote `b` to make easier and cleaner, and to reduce any possible performance impact I add a hint that it should be the same as the manual `a` calls.

But alas, if you so insist I'll add a comment linking to this discussion.

Reply:
https://github.com/hyperledger/solang/pull/1237/files#r1151621676

Reply:
@DaniPopes Inlining functions is not guaranteed to improve performance. It might also make the code slower:
https://isocpp.org/wiki/faq/inline-functions#inline-and-perf